presto-main/src/main/java/com/facebook/presto/operator/SpatialIndexBuilderOperator.java

```diff
@@ -0,0 +1,233 @@
+            PagesSpatialIndexFactory pagesSpatialIndexFactory,
+            List<Integer> outputChannels,
+            int indexChannel,
+            BiPredicate<OGCGeometry, OGCGeometry> spatialRelationshipTest,
+            Optional<JoinFilterFunctionFactory> filterFunctionFactory,
+            int expectedPositions,
+            PagesIndex.Factory pagesIndexFactory)
+    {
+        this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
+        this.localUserMemoryContext = operatorContext.localUserMemoryContext();
+        this.spatialRelationshipTest = requireNonNull(spatialRelationshipTest, "spatialRelationshipTest is null");
+        this.filterFunctionFactory = filterFunctionFactory;
+
+        this.pagesSpatialIndexFactory = requireNonNull(pagesSpatialIndexFactory, "pagesSpatialIndexFactory is null");
+        this.index = pagesIndexFactory.newPagesIndex(pagesSpatialIndexFactory.getTypes(), expectedPositions);
+
+        this.outputChannels = requireNonNull(outputChannels, "outputChannels is null");
+        this.indexChannel = indexChannel;
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public List<Type> getTypes()
+    {
+        return pagesSpatialIndexFactory.getTypes();
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finished;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        requireNonNull(page, "page is null");
+        checkState(!isFinished(), "Operator is already finished");
+
+        index.addPage(page);
+
+        if (!localUserMemoryContext.trySetBytes((index.getEstimatedSize().toBytes()))) {
+            index.compact();
+            localUserMemoryContext.setBytes(index.getEstimatedSize().toBytes());
+        }
+
+        operatorContext.recordGeneratedOutput(page.getSizeInBytes(), page.getPositionCount());
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        return null;
+    }
+
+    @Override
+    public ListenableFuture<?> isBlocked()
+    {
+        if (indexNotNeeded != null && !indexNotNeeded.isDone()) {
+            return indexNotNeeded;
+        }
+        return NOT_BLOCKED;
+    }
+
+    @Override
+    public void finish()
+    {
+        if (finishing) {
+            return;
+        }
+
+        finishing = true;
+        PagesSpatialIndexSupplier spatialIndex = index.createPagesSpatialIndex(operatorContext.getSession(), indexChannel, spatialRelationshipTest, filterFunctionFactory, outputChannels);
+        localUserMemoryContext.setBytes(spatialIndex.getEstimatedMemorySizeInBytes());
+        indexNotNeeded = pagesSpatialIndexFactory.lendPagesSpatialIndex(spatialIndex);
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        if (finished) {
+            return true;
+        }
+
+        if (finishing && indexNotNeeded.isDone()) {
+            index.clear();
```

Comment:
Add `close()` method and have `index.clear()` there too.
Like in https://github.com/prestodb/presto/pull/10039

Reply:
You can for example move 
```
index.clear();
localUserMemoryContext.setBytes(index.getEstimatedSize().toBytes());
```
to `close()` and just call it here.